### DISCUSSION TITLE
Command to raise GoNeoVim window

### DISCUSSION THREAD

**Discussion by gi1242** [akiyosi/goneovim#394] [Ideas]
Is there a way to raise the GoNeoVim window from a script? If no, would you consider providing one?

Use case: LaTeX inverse searches. Many plugins provide the functionality by which when the user clicks a PDF file, NeoVim (or whatever frontend) moves to the corresponding location in the source (LaTeX) file. It would be desirable to also raise the corresponding NeoVim Window, but this doesn't seem possible as far as I can tell. (Note, I'm using Wayland, so I can't raise windows using `xdotool` or the likes.)

If you could provide a function, say `GoneovimRaise()` (or a command) that raises the window, then it would improve the user experience a lot.

There are many probably other use cases.

Thanks in advance!

**Comment by akiyosi** [comment]
@gi1242 
Hi :) 
I think it is possible to create an API like `GoneovimRaise` to raise the Goneovim application window. (Note that the "Window" to raise is the application window, not the Neovim buffer display window.)
However, I am not yet aware of any specific usage scenario.
By what means are you thinking of invoking `GoneovimRaise` when you are viewing a PDF viewer and the window focus is also on the PDF viewer?

**Reply by gi1242** [reply]
Hey,

Thanks for your quick response!!

I should have explained my usage scenario better: [LaTeX](https://www.latex-project.org/) is a document preparation system that is widely used in the scientific community to write papers / technical documentation. It takes a source document (plain text file, typically extension `.tex`), and compiles it to a PDF. I (and many others) edit the source file in Vim/NeoVim/Goneovim/etc. There are excellent Vim frameworks that help editing, and compiling LaTeX files (look at [VimTex](https://github.com/lervag/vimtex), or [TexLab](https://github.com/fannheyward/coc-texlab) for instance.

For long documents (20 pages+), it's very useful to be able to go back and forth between the TeX file, and the PDF file. While editing the source TeX file, I press a key and the PDF viewer highlights the corresponding location in the PDF file. This is called a *forward search*, and is typically done using [SyncTeX](http://itexmac.sourceforge.net/SyncTeX.html).

The use case I described in my original post is for an *inverse search*. While viewing the PDF file, if you want to make a change, you typically Shift Click / Control Click in the PDF viewer, and it goes to the corresponding place in the TeX file. There are many Vim/NeoVim plugins that bring the Vim buffer to the foreground, move the cursor to the right place, and highlight the corresponding line.

**However** (and this is why I posted), none of them can raise the Vim window! Typically one uses some hack (xdotool, or for console Vims finding the PID of the parent terminal and then xdotool / etc.). These hacks aren't possible in Wayland. xdotool won't work by design.

So if the Vim front end (GoNeoVim) provides a way for Vim to tell it to raise the window, that would make things much more convenient for users doing LaTeX inverse searches. A command `:GoneovimRaise`, or a function `call GoneovimRaise()` would allow the script to position the cursor in the right place, and then also raise the Vim window.

Hopefully the above shows that there's a good use case, that would be of interest to a large number of users, and you're willing to consider adding it!

Thanks again.

GI

**Comment by akiyosi** [comment]
@gi1242 
Thank you for your detailed explanation.
Sorry for the delay in replying again.

I may not have fully understood it yet, but I first created a version with a simple command to raise the application window.
I don't know if this will fit your use case, so I would really appreciate your feedback. You can get the built artifacts from the following CI job

https://github.com/akiyosi/goneovim/actions/runs/2813098860

**Reply by akiyosi** [reply]
the added command is `GonvimWindowRaise`

**Reply by gi1242** [reply]
Thanks a lot! I just tested it on Plasma/Wayland (with focus stealing off). It doesn't work. It might be an issue with Plasma/Wayland (which unfortunately I have to use for other reasons). I'll see if there's a way to do it on Plasma/Wayland

**Reply by akiyosi** [reply]
Thanks for the feedback.
The implementation uses the general raise() function of the Qt framework. I will look into why it doesn't work with Plasma/Wayland here as well.

**Reply by akiyosi** [reply]
I have fixed the command process.
Could you please check the operation here?
https://github.com/akiyosi/goneovim/actions/runs/2832029269

**Comment by gi1242** [comment]
On Wed, Aug 10, 2022 at 06:57:53AM -0700, akiyosi wrote:

> I have fixed the command process.
> Could you please check the operation here?
> https://github.com/akiyosi/goneovim/actions/runs/2832029269

WOW! It works! Thanks so much. How did you do it?

Everything I googled seemed to say that it's not possible. Neovim-qt has
a function `GuiForeground()`, which does nothing on Plasma/Wayland.

Thanks again!

GI

-- 
Don't have awk? Use this simple sh emulation:
#!/bin/sh
echo 'Awk bailing out!'
exit 2

**Reply by akiyosi** [reply]
It incorporated all the Qt native functions related to window activation. 
Because of this, it is not certain which one was actually effective in your environment :P

https://github.com/akiyosi/goneovim/commit/11fb296b883feb3ecc90bccbe2f728bac1f60df8

**Comment by gi1242** [comment]
I just took a look at your code. I think your addition of

    editor.window.SetWindowState(core.Qt__WindowActive)

is what did the trick. Now I have to see if I can add this line of code
into my PDF viewer. (Unfortunately it is Gtk based, so probably it won't
work 🙄)

Thanks again!

GI

-- 
The energizer bunny went to jail. He was charged with battery.

**Comment by gi1242** [comment]
Oh bummer. It doesn't work 🙁.

So the issue is that when I tested it earlier, there was something wrong with the nightly build. It ran under XWayland, and not natively under Wayland.

I just recompiled it on my system (merged window-raise-cmd into master), so that it runs natively under Wayland. In this case the window raise command doesn't work 🙁. It prints out a message saying:

    qt.qpa.wayland: Wayland does not support QWindow::requestActivate()

I'll check again when the QT Wayland supports this... (It's blurry if I run it under XWayland, so unfortunately I can't run it under XWayland.)

Thanks again for adding it.

**Comment by akiyosi** [comment]
Thanks for the info.
I understand the situation.

By the way, I am concerned about the blurring issue(#399).
A similar problem has been reported in the most recent issue, so if you have any information you can provide as well, I would appreciate it.

**Reply by gi1242** [reply]
Unfortunately I don't think there's anything that can be done. The blurring is a Plasma/Wayland issue when using scaling, and running XWayland apps. There are a lot of quirks... but unfortunately using Plasma/X11 isn't really feasible when you have multiple displays with different resolutions. I'm waiting (impatiently :joy:) for it to be fixed ...

If there's some specific thing you want me to try on my system just let me konw. I use it  all the time...

**Reply by akiyosi** [reply]
Thanks for the comment!
Please let the discussion continue if there is something more.
